idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/WrapWithSafeLetCall.kt

```diff
@@ -0,0 +1,277 @@
+        if (expression == null) return null
+        var current = expression.typableParentExpression
+        while (current != null && !isExpressionAtNullablePosition(current)) {
+            current = current.typableParentExpression
+        }
+        return current
+    }
+
+    private fun KtAnalysisSession.isExpressionAtNullablePosition(expression: KtExpression): Boolean {
+        val parent = expression.parent
+        return when {
+            parent is KtProperty && expression == parent.initializer -> {
+                if (parent.typeReference == null) return true
+                val symbol = parent.getSymbol()
+                (symbol as KtTypedSymbol).annotatedType.type.isMarkedNullable
+            }
+            parent is KtValueArgument && expression == parent.getArgumentExpression() -> {
+                // In the following logic, if call is missing, unresolved, or contains error, we just stop here so the wrapped call would be
+                // inserted here.
+                val functionCall = parent.getParentOfType<KtCallExpression>(strict = true) ?: return true
+                val resolvedCall = functionCall.resolveCall() ?: return true
+                return isFunctionAcceptNull(resolvedCall, parent.argumentIndex) ?: true
+            }
+            parent is KtBinaryExpression -> {
+                if (parent.operationToken in KtTokens.ALL_ASSIGNMENTS && parent.left == expression) {
+                    // If current expression is an l-value in an assignment, just keep going up because one cannot assign to a let call.
+                    return false
+                }
+                val resolvedCall = parent.resolveCall()
+                when {
+                    resolvedCall != null -> {
+                        // The binary expression is a call to some function
+                        val isInExpression = parent.isInCall()
+                        val expressionIsArg = when {
+                            parent.left == expression -> isInExpression
+                            parent.right == expression -> !isInExpression
+                            else -> return true
+                        }
+                        isFunctionAcceptNull(resolvedCall, if (expressionIsArg) 0 else -1) ?: true
+                    }
+                    parent.operationToken == KtTokens.EQ -> {
+                        // The binary expression is a variable assignment
+                        parent.left?.getKtType()?.isMarkedNullable ?: true
+                    }
+                    // The binary expression is some unrecognized constructs so we stop here.
+                    else -> true
+                }
+            }
+            parent is KtContainerNodeForControlStructureBody || parent is KtBlockExpression -> true
+            else -> false
+        }
+    }
+
+    private val Name.identifierOrNullIfSpecial get() = takeIf { !it.isSpecial }?.identifier
```

Comment:
Oh, I always checks if `Name` is special and retrieve `identifier` if not. I think it's worth adding this to `Name` !

Reply:
Sounds good!